app/controllers/submissions_controller.rb

```diff
@@ -169,14 +177,16 @@ def run
     close_client_connection(client_socket)
     Rails.logger.debug { "Running a submission timed out: #{e.message}" }
     @output = "timeout: #{@output}"
+    @status = :timeout
     extract_durations(e)
   rescue Runner::Error => e
     client_socket.send_data JSON.dump({cmd: :status, status: :container_depleted})
     close_client_connection(client_socket)
     Rails.logger.debug { "Runner error while running a submission: #{e.message}" }
+    @status = :container_depleted
```

Comment:
Why is here no `@passed = false` like for the other `:container_depleted`? Is `@passed` a test-specific field and `@exit_code` a run-specific field (like it is used in `migrate_messages`)?

Reply:
Yes, exactly. In the `status` column, we store general information about the execution: Did it succeed and exited normally (status code 0)? Or did it fail? And if it failed, why (`failed` = user's fault, `timeout` = we cancelled, `out_of_memory` = OOM killer kicked in, ...). 

Additionally, we store the `exit_code`. That might be somehow duplicated with the `status` column: 
- `status = :ok` -> `exit_code = 0`
- `status = :out_of_memory` -> `exit_code = 137`
- `status = :timeout` -> no exit code
- `status = :container_depleted` -> no exit code
- `status = :failed` -> `exit_code >= 1 && exit_code <= 255 && exit_code != 137`

Additionally, we have the `passed` column. That is solely used for test executions, not for normal run executions. In that column, we store whether the test passed (in which case we don't store the output) or whether the test failed (in which case we have the output). It can be inferred from the status column as well:

- `status = :ok` -> `passed = true`
- `status = :failed` -> `passed = false`

I also considered reducing the number of columns, but I would still argue that some information are still valuable. Most probably, the `passed` column could be removed as I would be able to reconstruct it with the `cause` and `status` columns. Maybe that's something for another migration ;)